core/pta/rtc.c

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2018, Linaro Limited
+ * Copyright (c) 2021, EPAM Systems. All rights reserved.
+ * Copyright (c) 2022, Microchip
+ *
+ * Based on pta/hwrng.c
+ *
+ */
+
+#include <drivers/rtc.h>
+#include <kernel/pseudo_ta.h>
+#include <rtc_pta_client.h>
+#include <tee_api_defines.h>
+
+#define PTA_NAME "rtc.pta"
+
+static TEE_Result rtc_pta_get_time(uint32_t types,
+				   TEE_Param params[TEE_NUM_PARAMS])
+{
+	struct optee_rtc_time *time = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	time = params[0].memref.buffer;
+	if (!time || params[0].memref.size != sizeof(*time))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	rtc_get_time(time);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result rtc_pta_set_time(uint32_t types,
+				   TEE_Param params[TEE_NUM_PARAMS])
+{
+	struct optee_rtc_time *time = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
```

Comment:
We are transmitting 7 32-bit fields for struct optee_rtc_time. We can pass those fields as value parameters instead since we have 8 such fields available. With this the kernel RTC driver wouldn't need to use any shared memory for communication and here we wouldn't have to worry about the passed parameters being changed while used.

Reply:
Hmm, that's unfortunate. Then I guess we need some struct definition, but I'd rather have a separate one so internals of OP-TEE doesn't become part of a TA ABI.

Reply:
Agreed, I did the same on Linux side to avoid reusing the `struct rtc_time`. I can probably move this definition in `rtc_pta_client.h` and add another one in rtc.h. They will be identical though.